python-package/xgboost/spark/data.py

```diff
@@ -1,192 +1,166 @@
-# type: ignore
-"""Xgboost pyspark integration submodule for data related functions."""
-# pylint: disable=too-many-arguments
-from typing import Iterator
+from typing import Iterator, List, Sequence, Optional, Dict, Tuple, Callable
+from collections import defaultdict, namedtuple
 import numpy as np
 import pandas as pd
-from xgboost import DMatrix
-
-
-def _prepare_train_val_data(
-    data_iterator, has_weight, has_validation, has_fit_base_margin
-):
-    def gen_data_pdf():
-        for pdf in data_iterator:
-            yield pdf
-
-    return _process_data_iter(
-        gen_data_pdf(),
-        train=True,
-        has_weight=has_weight,
-        has_validation=has_validation,
-        has_fit_base_margin=has_fit_base_margin,
-        has_predict_base_margin=False,
-    )
-
-
-def _check_feature_dims(num_dims, expected_dims):
-    """
-    Check all feature vectors has the same dimension
-    """
-    if expected_dims is None:
-        return num_dims
-    if num_dims != expected_dims:
-        raise ValueError(
-            f"Rows contain different feature dimensions: Expecting {expected_dims}, got {num_dims}."
-        )
-    return expected_dims
-
-
-def _row_tuple_list_to_feature_matrix_y_w(
-    data_iterator,
-    train,
-    has_weight,
-    has_fit_base_margin,
-    has_predict_base_margin,
-    has_validation: bool = False,
-):
-    """
-    Construct a feature matrix in ndarray format, label array y and weight array w
-    from the row_tuple_list.
-    If train == False, y and w will be None.
-    If has_weight == False, w will be None.
-    If has_base_margin == False, b_m will be None.
-    Note: the row_tuple_list will be cleared during
-    executing for reducing peak memory consumption
-    """
-    # pylint: disable=too-many-locals
-    expected_feature_dims = None
-    label_list, weight_list, base_margin_list = [], [], []
-    label_val_list, weight_val_list, base_margin_val_list = [], [], []
-    values_list, values_val_list = [], []
-
-    # Process rows
-    for pdf in data_iterator:
-        if len(pdf) == 0:
-            continue
-        if train and has_validation:
-            pdf_val = pdf.loc[pdf["validationIndicator"], :]
-            pdf = pdf.loc[~pdf["validationIndicator"], :]
-
-        num_feature_dims = len(pdf["values"].values[0])
-
-        expected_feature_dims = _check_feature_dims(
-            num_feature_dims, expected_feature_dims
-        )
+from xgboost import DMatrix, DataIter, DeviceQuantileDMatrix
+from xgboost.compat import concat, lazy_isinstance
+
+
+def stack_df(df: pd.DataFrame) -> np.ndarray:
```

Comment:
May I ask what this function does?

Reply:
So, that is to say,  the ETL converts the single columns into column arrays, and then pandas_udf converts the column of arrays back into a single array? It looks like this operation is redundant and causes the performance issue?

By the way, does that "converts a column of arrays into a single array" cause memory copy?

Reply:
Yup, that's just the convention of spark I guess, a way to handle `feature_col`. We can support `feature_cols` later and recommend it in our document. But still, we need to handle the `feature_col` gracefully.

Reply:
If `stack_df` input is `pd.DataFrame`, that means the features column  in train dataset must be a spark struct type. @trivialfis Do we plan to support spark struct type as features input ?

Reply:
Seems we should use `array = s.to_numpy()` instead of `array = s.values`. WDYT ?